src/hooks/UrqlUseSubscription.re

```diff
@@ -81,23 +83,17 @@ let useSubscription =
       (),
     );
 
-  React.useMemo3(
-    () => {
-      let response: UrqlTypes.hookResponse(ret, 'extensions) =
-        switch (handler) {
-        | Handler(handlerFn) =>
-          useSubscriptionJs(
-            args,
-            Some((acc, data) => handlerFn(acc, parse(data))),
-          )[0]
-          |> useSubscriptionResponseToRecord(x => x)
-        | NoHandler =>
-          useSubscriptionJs(args, None)[0]
-          |> useSubscriptionResponseToRecord(parse)
-        };
+  let handler' =
+    switch (handler) {
+    | Handler(handlerFn) =>
+      Some((acc, data) => handlerFn(acc, parse(data)))
```

Comment:
I opted to always have the data coming out of `useSubscriptionJs` be consistent by passing a handler that will parse by default. However, I had to `Obj.magic` here because the types seemed to enforce that `acc` and `resp` be different. I started tracking it down but things didn't makes sense. `type handler` takes `'ret` but is not used?

Reply:
Could you elaborate a bit on why you opted that way? Is it to not pass `parse` as argument to `jsSubscriptionResponseToRecord`? Which benefits do you see?

The reason why the compiler expects resp and acc to be different is because of how `handler('acc, 'resp, 'ret)` has been declared (It's a [GADT](https://sketch.sh/s/yH0MJiujNSiofDWOU85loX/))

So what's happening here is that in the first branch `'acc` and `'ret` are the same, guaranteed by `handler('acc, 'resp, 'ret)` type, but for the second branch no relation is guaranteed between `'acc` and `'ret`

The type `'ret` in `handler('acc, 'resp, 'ret)` is there only to help the compiler know what's the return value for the `useSubscription` when passing a `Handler` and when passing a `NoHandler`

GADTs can be a little confusing, if you wish to, I'll hang out on Spectrum where we can chat in real time

Reply:
> Could you elaborate a bit on why you opted that way? Is it to not pass parse as argument to jsSubscriptionResponseToRecord? Which benefits do you see?

I guess my reasoning is something like this:

1. In terms of reasoning about hooks, calling them inside of a switch is much like a conditional. We're calling the same one so it works, but as a matter of principle I'd like to try and compute the arguments once and only call the hook once so I don't fall into a trap.
2. The fact that previously the result of one call to `useSubscription` had to be parsed and one with a handler did not seemed odd to me. I'd like the return values to be used in the same way. I'd also like to be parsing in the same way, in the same place. If you think of `handler` more like a transform rather than a function to mash new data on to the previous value, this seemed reasonable. The transform for `NoHandler` is identity. Now we can parse the data going into the transform consistently for both cases.
3. Since we had to memoize the conversion to a record, having a single definition of the function was just simpler. (I also removed `use` from the function name since it is not actually a hook)

> So what's happening here is that in the first branch 'acc and 'ret are the same, guaranteed by handler('acc, 'resp, 'ret) type, but for the second branch no relation is guaranteed between 'acc and 'ret

Ah! I see what you're doing now. Thanks for that! I switched `NoHandler` to be `handler('resp, 'resp, 'resp)` (don't ask to infer the `'acc` type) and it works now. Pushing up a change. A lot of changes that maybe weren't necessary here, but made it easier for me to reason about. I'm totally open to changing it if you prefer a solution closer to how it was previously.

Reply:
Your changes are just fine, and I also agree with you that it made it easier to reason about! :)